test/test_type_promotion.py

```diff
@@ -0,0 +1,201 @@
+import torch
+import unittest
+
+from common_utils import TestCase, run_tests, load_tests
+
+# load_tests from common_utils is used to automatically filter tests for
+# sharding on sandcastle. This line silences flake warnings
+load_tests = load_tests
+
+class TestTypePromotion(TestCase):
+
+    def setUp(self):
+        super(TestTypePromotion, self).setUp()
+        torch.set_default_dtype(torch.float32)
+        self.device = 'cpu'
+
+    # In-place operations don't promote.
+    # `int+float -> float` but `int.add_(float)` is rejected as an error.
+    # Promoting inplace would require re-allocating and copying the memory of the
+    # tensor data, since element size could change.
+    def test_inplace(self):
+        int_tensor = torch.ones([4, 4, 4], dtype=torch.int32, device=self.device)
+
+        self.assertRaisesRegex(RuntimeError, "can't be cast to", lambda: int_tensor.add_(1.5))
+
+        expected = torch.ones([4, 4, 4], device=self.device)
+        expected = expected.to(torch.int32)
+
+        long_tensor = torch.ones([4, 4, 4], dtype=torch.int64, device=self.device)
+        int_tensor.add_(long_tensor)
+        int_tensor.add_(1)
+        three = expected + 2
+        self.assertEqual(int_tensor, three)
+        self.assertEqual(int_tensor.dtype, torch.int32)
+
+        uint8_tensor = torch.tensor([1, 1, 1], dtype=torch.uint8)
+        half_tensor = torch.tensor([1, 1, 1], dtype=torch.half)
+        # uint8 is unsigned, half is signed.
+        self.assertRaisesRegex(RuntimeError, "can't be cast to", lambda: uint8_tensor.add_(half_tensor))
+
+        # We allow demotion from signed to unsigned, unlike numpy, because we
```

Comment:
nit: can you reorganize this comment a bit.  I'm not sure if I'm reading this correctly, but according to my interpretation, you have two different ideas in the same sentence and and the same idea in two sentences.

I think what you are trying to say is:
```
We allow demotion from signed to unsigned, unlike numpy, because:
1) We don't want the performance penalty of inspecting scalar values
2) We don't want 'signed' to be considered a distinct category in promotion rules, because if it were, uint16_tensor +5 would result in a long tensor, which is not what we want.
```